File: Java/Machine Learning/XGBoost.java
/*Implementation of a XGBoost algorithm.
Author - Phalesh Kolpe*/

import java.util.ArrayList;
import java.util.List;

public class XGBoost {
    private List<Tree> trees;
    private double learningRate;
    private int numIterations;

    public XGBoost(double learningRate, int numIterations) {
        this.learningRate = learningRate;
        this.numIterations = numIterations;
        this.trees = new ArrayList<>();
    }

    public void train(List<double[]> features, double[] labels) {
        double[] residuals = labels.clone(); // Initialize residuals with labels
        for (int iteration = 0; iteration < numIterations; iteration++) {
            Tree tree = new Tree();
            tree.train(features, residuals);
            trees.add(tree);

            // Update residuals
            double[] predictions = tree.predict(features);
            for (int i = 0; i < residuals.length; i++) {
                residuals[i] = labels[i] - learningRate * predictions[i];
            }
        }
    }

    public double[] predict(List<double[]> features) {
        double[] predictions = new double[features.size()];
        for (Tree tree : trees) {
            double[] treePredictions = tree.predict(features);
            for (int i = 0; i < predictions.length; i++) {
                predictions[i] += learningRate * treePredictions[i];
            }
        }
        return predictions;
    }

    private static class Tree {
        private Node root;

        public void train(List<double[]> features, double[] residuals) {
            root = buildTree(features, residuals);
        }

            private Node buildTree(List<double[]> features, double[] residuals) {
        if (shouldStop(residuals)) {
            return createLeafNode(residuals);
        }

        Split split = findBestSplit(features, residuals);
        if (split == null) {
            return createLeafNode(residuals);
        }

        List<double[]> leftFeatures = new ArrayList<>();
        List<double[]> rightFeatures = new ArrayList<>();
        double[] leftResiduals = new double[features.size()];
        double[] rightResiduals = new double[features.size()];

        for (int i = 0; i < features.size(); i++) {
            double[] instance = features.get(i);
            if (isSplitSatisfied(instance, split)) {
                leftFeatures.add(instance);
                leftResiduals[i] = residuals[i];
            } else {
                rightFeatures.add(instance);
                rightResiduals[i] = residuals[i];
            }
        }

        Node node = new Node(split);
        node.leftChild = buildTree(leftFeatures, leftResiduals);
        node.rightChild = buildTree(rightFeatures, rightResiduals);

        return node;
    }



        private boolean shouldStop(double[] residuals) {
            // Stop if all residuals are zero or all are the same
            double firstResidual = residuals[0];
            for (double residual : residuals) {
                if (residual != firstResidual) {
                    return false;
                }
            }
            return true;
        }

        private Node createLeafNode(double[] residuals) {
            return new Node(residuals[0]);
        }

        private Split findBestSplit(List<double[]> features, double[] residuals) {
            Split bestSplit = null;
            double bestSplitScore = Double.MAX_VALUE;

            for (int i = 0; i < features.get(0).length; i++) {
                for (double threshold : findThresholds(features, i)) {
                    Split split = new Split(i, threshold);
                    double splitScore = calculateSplitScore(features, residuals, split);
                    if (splitScore < bestSplitScore) {
                        bestSplit = split;
                        bestSplitScore = splitScore;
                    }
                }
            }

            return bestSplit;
        }

        private double[] findThresholds(List<double[]> features, int featureIndex) {
            double[] thresholds = new double[features.size() - 1];
            List<double[]> sortedFeatures = new ArrayList<>(features);
            sortedFeatures.sort((a, b) -> Double.compare(a[featureIndex], b[featureIndex]));

            for (int i = 0; i < thresholds.length; i++) {
                thresholds[i] = (sortedFeatures.get(i)[featureIndex] + sortedFeatures.get(i + 1)[featureIndex]) / 2.0;
            }

            return thresholds;
        }

        private boolean isSplitSatisfied(double[] instance, Split split) {
            return instance[split.featureIndex] <= split.threshold;
        }

            private double calculateSplitScore(List<double[]> features, double[] residuals, Split split) {
        double[] leftResiduals = new double[features.size()];
        double[] rightResiduals = new double[features.size()];

        for (int i = 0; i < features.size(); i++) {
            double[] instance = features.get(i);
            if (isSplitSatisfied(instance, split)) {
                leftResiduals[i] = residuals[i];
            } else {
                rightResiduals[i] = residuals[i];
            }
        }

        double leftScore = calculateNodeScore(leftResiduals);
        double rightScore = calculateNodeScore(rightResiduals);
        return leftScore + rightScore;
    }


        private double calculateNodeScore(double[] residuals) {
            double sum = 0.0;
            for (double residual : residuals) {
                sum += Math.pow(residual, 2);
            }
            return sum;
        }

        public double[] predict(List<double[]> features) {
            double[] predictions = new double[features.size()];
            for (int i = 0; i < features.size(); i++) {
                double[] instance = features.get(i);
                predictions[i] = predictInstance(instance);
            }
            return predictions;
        }

        private double predictInstance(double[] instance) {
            Node node = root;
            while (!node.isLeafNode()) {
                if (isSplitSatisfied(instance, node.split)) {
                    node = node.leftChild;
                } else {
                    node = node.rightChild;
                }
            }
            return node.value;
        }
    }

    private static class Node {
        private Split split;
        private Node leftChild;
        private Node rightChild;
        private double value;

        public Node(Split split) {
            this.split = split;
        }

        public Node(double value) {
            this.value = value;
        }

        public boolean isLeafNode() {
            return leftChild == null && rightChild == null;
        }
    }

    private static class Split {
        private int featureIndex;
        private double threshold;

        public Split(int featureIndex, double threshold) {
            this.featureIndex = featureIndex;
            this.threshold = threshold;
        }
    }

    public static void main(String[] args) {
        // Sample input features and labels
        List<double[]> features = new ArrayList<>();
        features.add(new double[]{1.0, 2.0, 3.0});
        features.add(new double[]{4.0, 5.0, 6.0});
        features.add(new double[]{7.0, 8.0, 9.0});

        double[] labels = new double[]{10.0, 20.0, 30.0};

        // Create an instance of XGBoost and train the model
        XGBoost xgBoost = new XGBoost(0.1, 100);
        xgBoost.train(features, labels);

        // Sample input for prediction
        List<double[]> testFeatures = new ArrayList<>();
        testFeatures.add(new double[]{2.0, 3.0, 4.0});
        testFeatures.add(new double[]{5.0, 6.0, 7.0});

        // Perform predictions
        double[] predictions = xgBoost.predict(testFeatures);

        // Print the predictions
        System.out.println("Predictions:");
        for (double prediction : predictions) {
            System.out.println(prediction);
        }
    }

}


    

File: C/Machine Learning/XGboost.c
/*Implementation of a XGBoost algorithm.
Author - Phalesh Kolpe*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Define constants
#define MAX_ITERATIONS 100
#define LEARNING_RATE 0.1

// Define the XGBoost model structure
typedef struct {
    double* weights;
} XGBoostModel;

// Define the XGBoost training function
void trainXGBoost(double X[][2], double* y, int num_samples, int num_features, XGBoostModel* model) {
    // Initialize the model weights
    model->weights = (double*)malloc((num_features + 1) * sizeof(double));
    for (int i = 0; i <= num_features; i++) {
        model->weights[i] = 0.0;
    }

    // Iterate through the training data
    for (int iteration = 0; iteration < MAX_ITERATIONS; iteration++) {
        // Compute predictions
        double* predictions = (double*)malloc(num_samples * sizeof(double));
        for (int i = 0; i < num_samples; i++) {
            double prediction = model->weights[0];  // Bias term

            for (int j = 0; j < num_features; j++) {
                prediction += model->weights[j + 1] * X[i][j];
            }

            predictions[i] = prediction;
        }

        // Compute gradients and update weights
        double* gradients = (double*)malloc(num_samples * sizeof(double));
        for (int i = 0; i < num_samples; i++) {
            gradients[i] = 2.0 * (predictions[i] - y[i]);
        }

        model->weights[0] -= LEARNING_RATE * gradients[0];  // Update bias term

        for (int j = 0; j < num_features; j++) {
            double gradient_sum = 0.0;
            for (int i = 0; i < num_samples; i++) {
                gradient_sum += gradients[i] * X[i][j];
            }

            model->weights[j + 1] -= LEARNING_RATE * gradient_sum;
        }

        // Clean up memory
        free(predictions);
        free(gradients);
    }
}

// Example usage
int main() {
    // Define the training data
    double X_train[5][2] = {
        {1.0, 1.0},
        {2.0, 3.0},
        {4.0, 5.0},
        {6.0, 7.0},
        {8.0, 9.0}
    };
    double y_train[5] = {3.0, 7.0, 9.0, 13.0, 17.0};

    // Create the model
    XGBoostModel model;

    // Train the model
    trainXGBoost(X_train, y_train, 5, 2, &model);

    // Print the learned weights
    printf("Learned Weights:\n");
    for (int i = 0; i <= 2; i++) {
        printf("Weight %d: %lf\n", i, model.weights[i]);
    }

    // Clean up memory
    free(model.weights);

    return 0;
}
